File: my_trace_pkg_cpp/src/trace_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

class TracePublisher : public rclcpp::Node
{
public:
    TracePublisher() : Node("trace_publisher")
    {
        publisher_ = this->create_publisher<std_msgs::msg::String>("hello_world", 10);

        message_.data = "Hello, world!";

        timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&TracePublisher::publishMessage, this));
    }

private:
    void publishMessage()
    {
        publisher_->publish(message_);
    }

    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    std_msgs::msg::String message_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TracePublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: my_trace_pkg_py/my_trace_pkg_py/trace_publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
# Import OpenTelemetry libraries
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

resource = Resource(attributes={
    "service.name": "service"
})

trace.set_tracer_provider(TracerProvider(resource=resource))
tracer = trace.get_tracer(__name__)

otlp_exporter = OTLPSpanExporter(endpoint="http://localhost:4317", insecure=True)

span_processor = BatchSpanProcessor(otlp_exporter)

trace.get_tracer_provider().add_span_processor(span_processor)

class TracePublisher(Node):
    def __init__(self):
        super().__init__('trace_publisher')
        self.publisher_ = self.create_publisher(String, 'trace_topic', 10)
        self.timer_ = self.create_timer(1, self.publish_message)
        self.get_logger().info('Trace publisher node initialized')

    def publish_message(self):
        with tracer.start_as_current_span("publish_message"):
            msg = String()
            msg.data = 'hello world'
            with tracer.start_as_current_span("publish"):
                self.publisher_.publish(msg)
                self.get_logger().info('Published: %s' % msg.data)

def main(args=None):
    rclpy.init(args=args)
    trace_publisher = TracePublisher()
    rclpy.spin(trace_publisher)
    trace_publisher.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()